File: 0704-binary-search/0704-binary-search.java
class Solution {
    public int search(int[] nums, int target) {
        //Using recursion
        int start=0;
        int end=nums.length-1;
        return binarySearch(nums,target,start,end);
    }
    private int binarySearch(int[] array,int target,int start,int end){
        if(start>end) return -1;
        int middle=start+(end-start)/2;
        if(target==array[middle]) return middle;
        else if(target>array[middle]){
            return binarySearch(array,target,middle+1,end);
        }
        else{
            return binarySearch(array,target,start,middle-1);
        }
        /*
        Using just iteration/loop
        int left=0;
        int right=nums.length-1;
        while(left<=right){
            int middle=left+(right-left)/2;
            if(nums[middle]==target){
             return middle;
            }
            else if(nums[middle]>target){
                 right=middle-1;
            }
            else{
                left=middle+1;
            }
        }
        return -1;
        */
    }
}

File: 0704-binary-search/0704-binary-search.py
class Solution:
    def binarySearch(self,array,target,left,right)->int:
        if left>right:
            return -1
        middle=left+(right-left)//2
        if array[middle]==target:
            return middle
        elif array[middle]<target:
            return self.binarySearch(array,target,middle+1,right)
        else:
            return self.binarySearch(array,target,left,middle-1)
    def search(self, nums: List[int], target: int) -> int:
        left=0
        right=len(nums)-1
        return self.binarySearch(nums,target,left,right)